File: apps/ios/Optio/Core/UI/AgentLogView.swift
import SwiftUI

/// Renders a transcript of `AgentLogEntry` rows the way the web log viewer does:
/// assistant text as prose, tool calls as collapsible monospace blocks, thinking
/// dimmed, errors red. Used for task logs, job run logs, review logs, session chat
/// and persistent-agent turns alike so every surface reads the same.
struct AgentLogView: View {
    let entries: [AgentLogEntry]
    var autoScroll = true

    var body: some View {
        ScrollViewReader { proxy in
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 10) {
                    ForEach(Array(entries.enumerated()), id: \.offset) { idx, entry in
                        AgentLogRow(entry: entry).id(idx)
                    }
                }
                .padding()
            }
            .onChange(of: entries.count) { _, count in
                guard autoScroll, count > 0 else { return }
                withAnimation { proxy.scrollTo(count - 1, anchor: .bottom) }
            }
        }
    }
}

struct AgentLogRow: View {
    let entry: AgentLogEntry
    @State private var expanded = false

    var body: some View {
        switch entry.type {
        case .text where isUser:
            userBubble
        case .text:
            Text(LocalizedStringKey(entry.content))
                .font(.body)
                .textSelection(.enabled)
                .frame(maxWidth: .infinity, alignment: .leading)
        case .thinking:
            Text(entry.content)
                .font(.footnote)
                .italic()
                .foregroundStyle(.secondary)
                .frame(maxWidth: .infinity, alignment: .leading)
        case .toolUse, .toolResult:
            toolBlock
        case .error:
            Label(entry.content, systemImage: "xmark.octagon")
                .font(.monoFootnote)
                .foregroundStyle(Tone.danger.textStyle)
                .frame(maxWidth: .infinity, alignment: .leading)
        case .system, .info, .unknown:
            Text(entry.content)
                .font(.caption.monospaced())
                .foregroundStyle(.secondary)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
    }

    private var toolName: String {
        entry.metadata?["toolName"]?.stringValue ?? (entry.type == .toolUse ? "tool" : "result")
    }

    // Optional metadata a producer may attach (see `LocalTranscriptLog`):
    // `role: "user"` marks a prompt typed by the human; `summary` is a tool call's
    // one-line summary shown in the header (the body is then the full input);
    // `result` / `resultIsError` fold the tool's result under its call.
    private var isUser: Bool { entry.metadata?["role"]?.stringValue == "user" }
    private var summary: String? { entry.metadata?["summary"]?.stringValue }
    private var pairedResult: String? { entry.metadata?["result"]?.stringValue }
    private var isError: Bool {
        entry.metadata?["resultIsError"]?.boolValue == true || (entry.metadata?["exitCode"]?.intValue ?? 0) != 0
    }
    private var hasBody: Bool { !entry.content.isEmpty || pairedResult != nil }

    private var userBubble: some View {
        HStack(alignment: .top, spacing: 8) {
            Image(systemName: "person.fill")
                .font(.caption2)
                .foregroundStyle(AppTheme.accent)
                .frame(width: 22, height: 22)
                .background(AppTheme.accent.opacity(0.15), in: Circle())
            VStack(alignment: .leading, spacing: 3) {
                HStack(spacing: 6) {
                    Text("You").font(.caption.weight(.semibold)).foregroundStyle(AppTheme.accent)
                    if let time = Self.shortTime(entry.timestamp) {
                        Text(time).font(.caption2).foregroundStyle(.tertiary).monospacedDigit()
                    }
                }
                Text(entry.content)
                    .font(.callout)
                    .textSelection(.enabled)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(.fill.tertiary, in: Radius.cardShape)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateStyle = .none
        f.timeStyle = .short
        return f
    }()

    private static func shortTime(_ iso: String) -> String? {
        guard !iso.isEmpty else { return nil }
        let parser = ISO8601DateFormatter()
        parser.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = parser.date(from: iso) ?? ISO8601DateFormatter().date(from: iso)
        return date.map { timeFormatter.string(from: $0) }
    }

    @ViewBuilder
    private var toolBody: some View {
        VStack(alignment: .leading, spacing: 6) {
            if !entry.content.isEmpty {
                Text(entry.content)
                    .font(.caption.monospaced())
                    .textSelection(.enabled)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(8)
                    .background(.fill.tertiary, in: Radius.smallShape)
            }
            if let pairedResult {
                Text(pairedResult.isEmpty ? "(no output)" : pairedResult)
                    .font(.caption.monospaced())
                    .foregroundStyle(isError ? Tone.danger.textStyle : AnyShapeStyle(.secondary))
                    .textSelection(.enabled)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(8)
                    .background(.fill.quaternary, in: Radius.smallShape)
            }
        }
    }

    private var toolLabel: some View {
        HStack(spacing: 6) {
            Image(systemName: isError ? "exclamationmark.circle" : entry.type == .toolUse ? "wrench.and.screwdriver" : "arrow.turn.down.left")
                .foregroundStyle(isError ? Tone.danger.textStyle : AnyShapeStyle(.secondary))
            Text(toolName)
                .font(.caption.weight(.semibold).monospaced())
                .foregroundStyle(isError ? Tone.danger.textStyle : AnyShapeStyle(.primary))
            if let exit = entry.metadata?["exitCode"]?.intValue, exit != 0 {
                Text("exit \(exit)").font(.caption2).foregroundStyle(.red)
            }
            Spacer(minLength: 4)
            Text((summary ?? entry.content).prefix(80).replacingOccurrences(of: "\n", with: " "))
                .font(.caption2.monospaced())
                .foregroundStyle(.tertiary)
                .lineLimit(1)
        }
    }

    @ViewBuilder
    private var toolBlock: some View {
        if hasBody {
            DisclosureGroup(isExpanded: $expanded) { toolBody } label: { toolLabel }
                .tint(.secondary)
        } else {
            toolLabel.padding(.leading, 2)
        }
    }
}

/// Text field + send button pinned above the keyboard. `onSend` receives trimmed text.
struct ChatComposer: View {
    var placeholder = "Message"
    var disabled = false
    /// Raise the keyboard on appear (deep links with `compose=1`).
    var autofocus = false
    var onSend: (String) async -> Void
    @State private var text = ""
    @State private var sending = false
    @FocusState private var focused: Bool

    var body: some View {
        HStack(alignment: .bottom, spacing: Spacing.s) {
            TextField(placeholder, text: $text, axis: .vertical)
                .lineLimit(1...6)
                .textFieldStyle(.plain)
                .padding(.horizontal, 14)
                .padding(.vertical, 9)
                .background(.fill.tertiary, in: Radius.bubbleShape)
                .focused($focused)
                .disabled(disabled)
            Button {
                Task { await send() }
            } label: {
                Group {
                    if sending { ProgressView().tint(.white) } else { Image(systemName: "arrow.up").font(.body.weight(.semibold)) }
                }
                .frame(width: 36, height: 36)
                .foregroundStyle(.white)
                .background(canSend ? AppTheme.accent : Color(.tertiaryLabel), in: Circle())
            }
            .buttonStyle(.plain)
            .disabled(!canSend)
            .accessibilityLabel("Send")
        }
        .padding(.horizontal, Spacing.l)
        .padding(.vertical, Spacing.s)
        .background(composerBackground)
        .onAppear { if autofocus { focused = true } }
        .sensoryFeedback(.impact(flexibility: .soft), trigger: sending) { _, new in new }
    }

    private var canSend: Bool { !disabled && !sending && !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

    @ViewBuilder
    private var composerBackground: some View {
        if #available(iOS 26, *) {
            Rectangle().fill(.clear).glassEffect(.regular, in: Rectangle())
        } else {
            Rectangle().fill(.bar)
        }
    }

    private func send() async {
        let msg = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !msg.isEmpty else { return }
        sending = true
        text = ""
        await onSend(msg)
        sending = false
    }
}

File: apps/android/core/ui/src/main/kotlin/dev/optio/core/ui/log/AgentLogView.kt
package dev.optio.core.ui.log

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.KeyboardReturn
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.outlined.Build
import androidx.compose.material.icons.outlined.Dangerous
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import dev.optio.core.model.AgentLogEntry
import dev.optio.core.model.boolValue
import dev.optio.core.model.intValue
import dev.optio.core.model.stringValue
import dev.optio.core.ui.components.MarkdownText
import dev.optio.core.ui.format.LocalClock
import dev.optio.core.ui.format.isoInstant
import dev.optio.core.ui.theme.OptioTheme
import dev.optio.core.ui.theme.Radius
import dev.optio.core.ui.theme.Spacing
import dev.optio.core.ui.theme.italic
import dev.optio.core.ui.theme.mono
import dev.optio.core.ui.theme.semibold
import dev.optio.core.ui.theme.tabularNums
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

/**
 * A transcript of [AgentLogEntry] rows rendered the way the web log viewer does (iOS
 * `AgentLogView`): assistant text as Markdown prose, prompts you typed as a "You" card, thinking
 * dimmed, tool calls as collapsible mono blocks (with their paired result folded under them),
 * errors red, system lines small mono. Used for task logs, job-run logs, review logs, session chat
 * and persistent-agent turns alike, so every surface reads the same.
 *
 * With [autoScroll] the list opens at the newest entry and follows new entries while the reader is
 * at the bottom; scrolling up stops following until they return. [emptyContent] shows instead of
 * an empty list. Test tag: `agent-log`; rows `log-row-<index>`.
 */
@Composable
fun AgentLogView(
    entries: List<AgentLogEntry>,
    modifier: Modifier = Modifier,
    autoScroll: Boolean = true,
    state: LazyListState = rememberLazyListState(),
    contentPadding: PaddingValues = PaddingValues(Spacing.l),
    emptyContent: (@Composable () -> Unit)? = null,
) {
    if (entries.isEmpty() && emptyContent != null) {
        Box(modifier.fillMaxSize(), contentAlignment = Alignment.Center) { emptyContent() }
        return
    }
    var previousCount by remember { mutableIntStateOf(0) }
    LaunchedEffect(entries.size) {
        val count = entries.size
        val previous = previousCount
        previousCount = count
        if (!autoScroll || count == 0) return@LaunchedEffect
        // Follow when the reader could see the previous last entry (appending doesn't move the list).
        val lastVisible = state.layoutInfo.visibleItemsInfo.lastOrNull()?.index ?: -1
        if (previous == 0 || lastVisible >= previous - 1) state.scrollToBottom(animated = previous != 0)
    }
    LazyColumn(
        state = state,
        modifier = modifier.fillMaxSize().testTag("agent-log"),
        contentPadding = contentPadding,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        itemsIndexed(entries, key = { index, _ -> index }) { index, entry ->
            AgentLogRow(entry, Modifier.testTag("log-row-$index"))
        }
    }
}

/** Scrolls to the end of the last item (iOS `scrollTo(last, anchor: .bottom)`). */
private suspend fun LazyListState.scrollToBottom(animated: Boolean) {
    val last = layoutInfo.totalItemsCount - 1
    if (last < 0) return
    if (animated) animateScrollToItem(last) else scrollToItem(last)
    val item = layoutInfo.visibleItemsInfo.lastOrNull() ?: return
    val overflow = item.offset + item.size - (layoutInfo.viewportEndOffset - layoutInfo.afterContentPadding)
    if (overflow > 0) {
        if (animated) animateScrollToItem(last, overflow) else scrollToItem(last, overflow)
    }
}

/**
 * One transcript entry (iOS `AgentLogRow`). Optional metadata a producer may attach: `role: "user"`
 * marks a prompt typed by the human; `toolName` names a tool call; `summary` is a call's one-line
 * summary for the header (the body is then the full input); `result` / `resultIsError` fold the
 * tool's result under its call; a non-zero `exitCode` marks a failed command.
 */
@Composable
fun AgentLogRow(
    entry: AgentLogEntry,
    modifier: Modifier = Modifier,
) {
    val colors = OptioTheme.colors
    val type = OptioTheme.type
    val meta = entry.metadata
    when (entry.type) {
        AgentLogEntry.TypeValue.TEXT -> if (meta?.get("role")?.stringValue == "user") {
            UserPrompt(entry, modifier)
        } else {
            SelectionContainer(modifier.fillMaxWidth()) { MarkdownText(entry.content) }
        }
        AgentLogEntry.TypeValue.THINKING -> SelectionContainer(modifier.fillMaxWidth()) {
            Text(entry.content, style = type.footnote.italic(), color = colors.secondaryLabel)
        }
        AgentLogEntry.TypeValue.TOOL_USE, AgentLogEntry.TypeValue.TOOL_RESULT -> ToolBlock(entry, modifier)
        AgentLogEntry.TypeValue.ERROR -> Row(
            modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Icon(Icons.Outlined.Dangerous, contentDescription = "Error", tint = colors.red, modifier = Modifier.padding(top = 1.dp).size(16.dp))
            SelectionContainer { Text(entry.content, style = type.monoFootnote, color = colors.red) }
        }
        AgentLogEntry.TypeValue.SYSTEM, AgentLogEntry.TypeValue.INFO, AgentLogEntry.TypeValue.UNKNOWN -> SelectionContainer(modifier.fillMaxWidth()) {
            Text(entry.content, style = type.caption.mono(), color = colors.secondaryLabel)
        }
    }
}

@Composable
private fun UserPrompt(entry: AgentLogEntry, modifier: Modifier) {
    val colors = OptioTheme.colors
    val type = OptioTheme.type
    Row(modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(Spacing.s)) {
        Box(
            Modifier
                .size(22.dp)
                .background(colors.accent.copy(alpha = 0.15f), CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Icon(Icons.Filled.Person, contentDescription = null, tint = colors.accent, modifier = Modifier.size(13.dp))
        }
        Column(
            Modifier
                .weight(1f)
                .background(colors.fillTertiary, Radius.cardShape)
                .padding(horizontal = Spacing.m, vertical = Spacing.s),
            verticalArrangement = Arrangement.spacedBy(3.dp),
        ) {
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                Text("You", style = type.caption.semibold(), color = colors.accent)
                shortTime(entry.timestamp, LocalClock.current.zone)?.let { Text(it, style = type.caption2.tabularNums(), color = colors.tertiaryLabel) }
            }
            SelectionContainer { Text(entry.content, style = type.callout, color = colors.label) }
        }
    }
}

@Composable
private fun ToolBlock(entry: AgentLogEntry, modifier: Modifier) {
    val colors = OptioTheme.colors
    val type = OptioTheme.type
    val meta = entry.metadata
    val isUse = entry.type == AgentLogEntry.TypeValue.TOOL_USE
    val toolName = meta?.get("toolName")?.stringValue ?: if (isUse) "tool" else "result"
    val summary = meta?.get("summary")?.stringValue
    val pairedResult = meta?.get("result")?.stringValue
    val exitCode = meta?.get("exitCode")?.intValue
    val isError = meta?.get("resultIsError")?.boolValue == true || (exitCode ?: 0) != 0
    val hasBody = entry.content.isNotEmpty() || pairedResult != null
    var expanded by rememberSaveable { mutableStateOf(false) }
    val chevron by animateFloatAsState(if (expanded) 180f else 0f, label = "chevron")

    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        val header = if (hasBody) {
            Modifier
                .clickable(role = Role.Button, onClickLabel = if (expanded) "Collapse" else "Expand") { expanded = !expanded }
                .semantics { stateDescription = if (expanded) "Expanded" else "Collapsed" }
        } else {
            Modifier.padding(start = 2.dp)
        }
        Row(
            header.fillMaxWidth().padding(vertical = 2.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Icon(
                when {
                    isError -> Icons.Outlined.ErrorOutline
                    isUse -> Icons.Outlined.Build
                    else -> Icons.AutoMirrored.Outlined.KeyboardReturn
                },
                contentDescription = null,
                tint = if (isError) colors.red else colors.secondaryLabel,
                modifier = Modifier.size(15.dp),
            )
            Text(
                toolName,
                style = type.caption.semibold().mono(),
                color = if (isError) colors.red else colors.label,
                maxLines = 1,
            )
            if (exitCode != null && exitCode != 0) {
                Text("exit $exitCode", style = type.caption2, color = colors.red, maxLines = 1)
            }
            Text(
                (summary ?: entry.content).take(80).replace('\n', ' '),
                style = type.caption2.mono(),
                color = colors.tertiaryLabel,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f).padding(start = 2.dp),
            )
            if (hasBody) {
                Icon(
                    Icons.Filled.ExpandMore,
                    contentDescription = null,
                    tint = colors.tertiaryLabel,
                    modifier = Modifier.size(18.dp).rotate(chevron),
                )
            }
        }
        AnimatedVisibility(expanded && hasBody, enter = fadeIn() + expandVertically(), exit = fadeOut() + shrinkVertically()) {
            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                if (entry.content.isNotEmpty()) {
                    SelectionContainer {
                        Text(
                            entry.content,
                            style = type.caption.mono(),
                            color = colors.label,
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(colors.fillTertiary, Radius.smallShape)
                                .padding(Spacing.s),
                        )
                    }
                }
                if (pairedResult != null) {
                    SelectionContainer {
                        Text(
                            pairedResult.ifEmpty { "(no output)" },
                            style = type.caption.mono(),
                            color = if (isError) colors.red else colors.secondaryLabel,
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(colors.fillQuaternary, Radius.smallShape)
                                .padding(Spacing.s),
                        )
                    }
                }
            }
        }
    }
}

private val shortTimeFormatter: DateTimeFormatter = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)

/** "4:02 PM" in the device zone for an ISO timestamp; null when empty or unparseable. */
internal fun shortTime(iso: String, zone: ZoneId = ZoneId.systemDefault()): String? {
    if (iso.isEmpty()) return null
    return iso.isoInstant()?.atZone(zone)?.let(shortTimeFormatter::format)
}
